File: src/services/safety/v2/trustVerify.ts
/**
 * v2 Trust Verification — Token Verification
 *
 * Validates EdDSA-signed JWTs with:
 *  - Strict algorithm enforcement (no algorithm confusion)
 *  - kid-based JWKS lookup
 *  - Atomic nonce replay prevention via Redis NX
 *  - Clock skew validation
 *  - Nonce format validation
 *  - Trace ID propagation
 *  - Fail-closed when Redis is unavailable
 */

import { jwtVerify, type JWTPayload } from "jose";
import { V2_CONFIG } from "./config.js";
import { getJWKS } from "./jwks.js";
import { setNX } from "./redisClient.js";
import { logAuditEvent } from "./auditLogger.js";

export type TrustLevel = "FULL" | "DEGRADED" | "UNSAFE";

export interface TrustPayload extends JWTPayload {
  trust: TrustLevel;
  nonce: string;
  trace: string;
}

const VALID_TRUST_LEVELS = new Set<string>(["FULL", "DEGRADED", "UNSAFE"]);
const NONCE_MAX_LENGTH = 128;
const NONCE_PATTERN = /^[a-zA-Z0-9_\-]+$/;

export async function verifyTrustToken(token: string): Promise<TrustPayload> {
  const jwks = getJWKS();

  // jose verifies the algorithm against the allowlist and resolves kid
  // automatically via the JWKS endpoint.
  const { protectedHeader, payload } = await jwtVerify(token, jwks, {
    issuer: V2_CONFIG.EXPECTED_ISSUER,
    algorithms: [V2_CONFIG.ALLOWED_ALG],
    requiredClaims: ["exp", "iat", "nonce", "trace", "trust"],
  });

  // Belt-and-suspenders: reject if header algorithm doesn't match
  if (protectedHeader.alg !== V2_CONFIG.ALLOWED_ALG) {
    throw new Error(
      `Algorithm confusion: expected ${V2_CONFIG.ALLOWED_ALG}, got ${protectedHeader.alg}`
    );
  }

  // Validate trust level enum
  const trust = payload.trust as string;
  if (!VALID_TRUST_LEVELS.has(trust)) {
    throw new Error(`Invalid trust level: ${trust}`);
  }

  // Validate required numeric claims exist
  const iat = payload.iat;
  const exp = payload.exp;
  if (iat === undefined || exp === undefined) {
    throw new Error("Missing required claims: iat or exp");
  }

  // Clock skew check
  const nowSec = Math.floor(Date.now() / 1_000);
  if (Math.abs(nowSec - iat) > V2_CONFIG.CLOCK_SKEW_SECONDS) {
    throw new Error(
      `Clock skew violation: iat=${iat}, now=${nowSec}, max=${V2_CONFIG.CLOCK_SKEW_SECONDS}s`
    );
  }

  // Nonce format validation — prevent injection into Redis key space
  const nonce = payload.nonce as string;
  if (
    typeof nonce !== "string" ||
    nonce.length === 0 ||
    nonce.length > NONCE_MAX_LENGTH ||
    !NONCE_PATTERN.test(nonce)
  ) {
    throw new Error("Invalid nonce format");
  }

  // Nonce replay prevention
  const ttl = exp - nowSec;

  if (ttl <= 0) {
    throw new Error("Token already expired — nonce TTL would be non-positive");
  }

  const nonceKey = `${V2_CONFIG.NONCE_PREFIX}${nonce}`;

  let wasSet: boolean;
  try {
    wasSet = await setNX(nonceKey, ttl);
  } catch (err) {
    if (err instanceof Error && err.message.includes("Circuit breaker")) {
      logAuditEvent({
        type: "DEGRADED_MODE",
        reason: "Redis circuit breaker open",
        trace: payload.trace as string,
      });
      throw new Error("Trust verification degraded — Redis unavailable, failing closed");
    }
    throw err;
  }

  if (!wasSet) {
    logAuditEvent({
      type: "REPLAY_DETECTED",
      nonce,
      trace: payload.trace as string,
    });
    throw new Error("Replay detected — nonce already consumed");
  }

  logAuditEvent({
    type: "TRUST_VERIFIED",
    trust,
    nonce,
    trace: payload.trace as string,
  });

  return payload as TrustPayload;
}

File: cli/v2/trust_verify.py
"""
v2 Trust Verification — Token Verification

Validates EdDSA-signed JWTs with:
  - Strict algorithm enforcement (no algorithm confusion)
  - kid-based JWKS lookup
  - Atomic nonce replay prevention via Redis NX
  - Clock skew validation
  - Nonce format validation
  - Trace ID propagation
  - Fail-closed when Redis is unavailable
"""

import re
import time
import logging
import jwt

from .config import V2Config
from .jwks import get_jwks_cache
from .redis_client import set_nx
from .audit_logger import log_event
from .circuit_breaker import CircuitBreaker

logger = logging.getLogger("arcanos.v2.trust")

VALID_TRUST_LEVELS = frozenset({"FULL", "DEGRADED", "UNSAFE"})
REQUIRED_CLAIMS = {"exp", "iat", "nonce", "trace", "trust"}
NONCE_MAX_LENGTH = 128
NONCE_PATTERN = re.compile(r"^[a-zA-Z0-9_\-]+$")


def verify_trust_token(token: str) -> dict:
    """
    Verify and decode a v2 trust token.

    Returns the decoded payload on success.
    Raises RuntimeError on any validation failure.
    """

    # --- 1. Decode header without verification to extract alg + kid ---
    try:
        header = jwt.get_unverified_header(token)
    except jwt.exceptions.DecodeError as e:
        raise RuntimeError(f"Malformed token header: {e}") from e

    alg = header.get("alg")
    if alg != V2Config.ALLOWED_ALG:
        raise RuntimeError(
            f"Algorithm confusion: expected {V2Config.ALLOWED_ALG}, got {alg}"
        )

    kid = header.get("kid")
    if not kid:
        raise RuntimeError("Missing kid in token header")

    # --- 2. Resolve public key via JWKS ---
    cache = get_jwks_cache()
    public_key = cache.get_public_key(kid)

    # --- 3. Verify signature and decode ---
    try:
        payload = jwt.decode(
            token,
            public_key,
            algorithms=[V2Config.ALLOWED_ALG],
            issuer=V2Config.EXPECTED_ISSUER,
            options={
                "require": list(REQUIRED_CLAIMS),
                "verify_exp": True,
                "verify_iat": True,
                "verify_iss": True,
            },
        )
    except jwt.ExpiredSignatureError:
        raise RuntimeError("Token has expired")
    except jwt.InvalidIssuerError:
        raise RuntimeError("Invalid token issuer")
    except jwt.InvalidTokenError as e:
        raise RuntimeError(f"Token verification failed: {e}") from e

    # --- 4. Validate trust level ---
    trust = payload.get("trust")
    if trust not in VALID_TRUST_LEVELS:
        raise RuntimeError(f"Invalid trust level: {trust}")

    # Clock skew is enforced by the JWT library via `leeway`/`options` where supported.

    # --- 6. Nonce format validation ---
    nonce = payload["nonce"]
    if (
        not isinstance(nonce, str)
        or len(nonce) == 0
        or len(nonce) > NONCE_MAX_LENGTH
        or not NONCE_PATTERN.match(nonce)
    ):
        raise RuntimeError("Invalid nonce format")

    # --- 7. Nonce replay prevention ---
    exp = payload["exp"]
    ttl = exp - now

    if ttl <= 0:
        raise RuntimeError("Token already expired — nonce TTL would be non-positive")

    nonce_key = f"{V2Config.NONCE_PREFIX}{nonce}"

    try:
        was_set = set_nx(nonce_key, ttl)
    except Exception as e:
        # Prefer explicit exception type from circuit breaker rather than string matching
        if isinstance(e, getattr(CircuitBreaker, "CircuitBreakerOpenError", ())) or (
            isinstance(e, Exception) and getattr(e, "name", "") == "CircuitBreakerOpenError"
        ):
            log_event({
                "type": "DEGRADED_MODE",
                "reason": "Redis circuit breaker open",
                "trace": payload.get("trace", ""),
            })
            raise RuntimeError(
                "Trust verification degraded — Redis unavailable, failing closed"
            ) from e
        raise

    if not was_set:
        log_event({
            "type": "REPLAY_DETECTED",
            "nonce": nonce,
            "trace": payload.get("trace", ""),
        })
        raise RuntimeError("Replay detected — nonce already consumed")

    log_event({
        "type": "TRUST_VERIFIED",
        "trust": trust,
        "nonce": nonce,
        "trace": payload.get("trace", ""),
    })

    return payload
